src/robot/running/keywordrunner.py

```diff
@@ -179,13 +188,54 @@ def _get_values_for_one_round(self, data):
 
     def _replace_variables(self, data):
         values = self._context.variables.replace_list(data.values)
-        if data.range:
-            values = self._get_range_items(values)
-        if len(values) % len(data.variables) == 0:
+        values = self._transform_items(values)
+        values_per_iteration = self._original_values_per_iteration(data.variables)
+        if len(values) % values_per_iteration == 0:
             return values
+        error_addendum = ""
+        if len(data.variables) != values_per_iteration:
+            # This is a bit awkward, but I'm not sure how else to phrase
+            # the impact of a non-value variable like from ForInEnumerate.
+            error_addendum = " (%d of which matter for this error)" % values_per_iteration
         raise DataError('Number of FOR loop values should be multiple of '
-                        'variables. Got %d variables but %d value%s.'
-                        % (len(data.variables), len(values), s(values)))
+                        'variables. Got %d variables%s but %d value%s.'
+                        % (len(data.variables), error_addendum,
+                            len(values), s(values)))
+
+    def _run_one_round(self, data, values):
+        name = ', '.join(format_assign_message(var, item)
+                         for var, item in zip(data.variables, values))
+        result = KeywordResult(kwname=name,
+                               type=data.FOR_ITEM_TYPE)
+        for var, value in zip(data.variables, values):
+            self._context.variables[var] = value
+        runner = KeywordRunner(self._context, self._templated)
+        with StatusReporter(self._context, result):
+            runner.run_keywords(data.keywords)
+
+    def _transform_items(self, items):
+        return items
+
+    def _original_values_per_iteration(self, values):
+        """
+        The number of values per iteration;
+        used to check if we have (a multiple of this) values.
+
+        This is its own method to support loops like ForInEnumerate
+        which add/remove items to the pool.
+        """
+        return len(values)
+
+
+class ForInRangeLoopRunner(BasicForLoopRunner):
```

Comment:
I'd like to drop `Loop` from the name and use `ForInRangeRunner`, `ForInZipRunner`, etc.